packages/typescript-estree/src/typedefs.ts

```diff
@@ -0,0 +1,1380 @@
+  | TSInferType
+  | TSInterfaceHeritage
+  | TSIntersectionType
+  | TSLiteralType
+  | TSMappedType
+  | TSNeverKeyword
+  | TSNullKeyword
+  | TSNumberKeyword
+  | TSObjectKeyword
+  | TSOptionalType
+  | TSParenthesizedType
+  | TSRestType
+  | TSStringKeyword
+  | TSSymbolKeyword
+  | TSThisType
+  | TSTupleType
+  | TSTypeLiteral
+  | TSTypeOperator
+  | TSTypeReference
+  | TSTypePredicate
+  | TSTypeQuery
+  | TSUndefinedKeyword
+  | TSUnionType
+  | TSUnknownKeyword
+  | TSVoidKeyword;
+export type TSUnaryExpression =
+  | AwaitExpression
+  | LeftHandSideExpression
+  | TSTypeAssertion
+  | UnaryExpression
+  | UpdateExpression;
+
+///////////////
+// Base, common types
+//  **Ensure you sort the interfaces alphabetically**
+///////////////
+
+interface BinaryExpressionBase extends NodeBase {
+  operator: string;
+  left: Expression;
+  right: Expression;
+}
+
+interface ClassDeclarationBase extends NodeBase {
+  typeParameters?: TSTypeParameterDeclaration;
+  superTypeParameters?: TSTypeParameterInstantiation;
+  id?: Identifier;
+  body: ClassBody;
+  superClass?: LeftHandSideExpression;
+  implements?: ExpressionWithTypeArguments[];
+  abstract?: boolean;
+  declare?: boolean;
+  decorators?: Decorator[];
+}
+
+interface ClassPropertyBase extends NodeBase {
+  key: PropertyName;
+  value: Expression;
+  computed: boolean;
+  static: boolean;
+  readonly?: boolean;
+  decorators?: Decorator[];
+  accessibility?: Accessibility;
+  optional?: boolean;
+  definite?: boolean;
+  typeAnnotation?: TSTypeAnnotation;
+}
+
+interface FunctionDeclarationBase extends NodeBase {
+  id: Identifier | null;
+  generator: boolean;
+  expression: boolean;
+  async: boolean;
+  params: Parameter[];
+  body?: BlockStatement | null;
+  returnType?: TSTypeAnnotation;
+  typeParameters?: TSTypeParameterDeclaration;
+}
+
+interface FunctionSignatureBase extends NodeBase {
+  params: Parameter[];
+  returnType?: TSTypeAnnotation;
+  typeParameters?: TSTypeParameterDeclaration;
+}
+
+interface LiteralBase extends NodeBase {
+  raw: boolean | number | RegExp | string | null;
+  value: string;
+  regex?: {
+    pattern: string;
+    flags: string;
+  };
+}
+
+interface MethodDefinitionBase extends NodeBase {
+  key: PropertyName;
+  value: FunctionExpression;
+  computed: boolean;
+  static: boolean;
+  kind: 'method' | 'get' | 'set' | 'constructor';
+  decorators?: Decorator[];
+  accessibility?: Accessibility;
+  typeParameters?: TSTypeParameterDeclaration;
+}
+
+interface TSHeritageBase extends NodeBase {
+  expression: Expression;
+  typeParameters?: TSTypeParameterDeclaration;
+}
+
+interface UnaryExpressionBase extends NodeBase {
+  operator: string;
+  prefix: boolean;
+  argument: LeftHandSideExpression | Literal | UnaryExpression;
+}
+
+///////////////
+// Typescript ESTree Nodes
+//  **Ensure you sort the interfaces alphabetically**
+///////////////
+
+export interface ArrayExpression extends NodeBase {
+  type: 'ArrayExpression';
```

Comment:
all of types should use `enum AST_NODE_TYPES`, it will ensure that names are set correctly
```ts
import { AST_NODE_TYPES } from './ast-node-types';
```

Reply:
and its used in `deeplyCopy()` function for check if we are supporting node
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/convert.ts#L283